include/ginkgo/core/log/batch_logger.hpp

```diff
@@ -0,0 +1,184 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright (c) 2017-2023, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_PUBLIC_CORE_LOG_BATCH_LOGGER_HPP_
+#define GKO_PUBLIC_CORE_LOG_BATCH_LOGGER_HPP_
+
+
+#include <memory>
+
+
+#include <ginkgo/core/base/batch_multi_vector.hpp>
+#include <ginkgo/core/base/types.hpp>
+#include <ginkgo/core/log/logger.hpp>
+
+
+namespace gko {
+namespace batch {
+/**
+ * @brief The logger namespace .
+ * @ref log
+ * @ingroup log
+ */
+namespace log {
+namespace detail {
+
+
+/**
+ * Stores logging data for batch solver kernels.
+ *
+ * @note Supports only single rhs
+ */
+template <typename ValueType>
+struct log_data final {
+    using real_type = remove_complex<ValueType>;
+
+    log_data(std::shared_ptr<const Executor> exec, size_type num_batch_items)
+        : res_norms(exec), iter_counts(exec)
+    {
+        const size_type workspace_size =
+            num_batch_items * (sizeof(real_type) + sizeof(int));
+        if (num_batch_items > 0) {
+            iter_counts.resize_and_reset(num_batch_items);
+            res_norms.resize_and_reset(num_batch_items);
+        } else {
+            GKO_INVALID_STATE("Invalid num batch items passed in");
+        }
+    }
+
+    log_data(std::shared_ptr<const Executor> exec, size_type num_batch_items,
+             array<unsigned char> workspace)
```

Comment:
it still has the issue when the workspace is not passed by array view.

Reply:
I think now when workspace is not a view, it should throw.